lib/atomic_lti/open_id_middleware.rb

```diff
@@ -146,46 +151,33 @@ def matches_target_link?(request)
     end
 
     def handle_lti_launch(env, request)
-      id_token = request.params["id_token"]
-      url = request.url
-
-      payload = valid_token(id_token: id_token, url: url)
-      if !payload
-        Rails.logger.info("Invalid lti launch: id_token: #{payload} - id_token: #{id_token} - url: #{url}")
-        return render_error(401, "Invalid LTI launch. Please launch the application again.")
-      end
+      id_token_decoded, state, state_verified = validate_launch(env, request, true)
 
-      # Validate the state and csrf token
-      state = request.params["state"]
-      csrf_token = request.cookies["open_id_#{state}"] || request.params["csrf_token"]
-      if csrf_token.blank? && AtomicLti.enforce_csrf_protection
-        return render_error(401, "Unauthorized. Please check that your browser allows cookies.")
-      end
-
-      if !AtomicLti::OpenId.validate_state(payload["nonce"], state, csrf_token)
-        return render_error(401, "Invalid launch state")
-      end
-
-      decoded_jwt = payload
-
-      update_install(id_token: decoded_jwt)
-      update_platform_instance(id_token: decoded_jwt)
-      update_deployment(id_token: decoded_jwt)
-      update_lti_context(id_token: decoded_jwt)
+      id_token = request.params["id_token"]
+      update_install(id_token: id_token_decoded)
+      update_platform_instance(id_token: id_token_decoded)
+      update_deployment(id_token: id_token_decoded)
+      update_lti_context(id_token: id_token_decoded)
 
-      errors = decoded_jwt.dig(AtomicLti::Definitions::TOOL_PLATFORM_CLAIM, "errors")
+      errors = id_token_decoded.dig(AtomicLti::Definitions::TOOL_PLATFORM_CLAIM, "errors")
       if errors.present? && !errors["errors"].empty?
         Rails.logger.error("Detected errors in lti launch: #{errors}, id_token: #{id_token}")
       end
 
-      env["atomic.validated.decoded_id_token"] = decoded_jwt
+      env["atomic.validated.decoded_id_token"] = id_token_decoded
       env["atomic.validated.id_token"] = id_token
 
+      #platform = AtomicLti::Platform.find_by!(iss: id_token["iss"])
+      platform = AtomicLti::Platform.find_by!(iss: id_token_decoded["iss"])
+      if request.params["lti_storage_target"].present? && AtomicLti.use_post_message_storage
+        lti_storage_params = build_lti_storage_params(request, platform)
+        # Add the values needed to do client side validate to the environment
+        env["atomic.validated.state_verified"] = state_verified
+        env["atomic.validated.state"] = state
+        env["atomic.validated.lti_storage_params"] = lti_storage_params
```

Comment:
We could wrap them all together:
```
 env["atomic.validated.state_validation"] = { state: state, lti_storage_params: lti_storage_params, verified_by_cookie: state_verified }
```
And then the render code in the app just has to make that one thing available.

Reply:
I changed it